sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+                        // unexpected happened and it stopped working midway, this is the moment we expect to catch an exception.
+
+                        try
+                        {
+                            await ActiveLoadBalancingTask.ConfigureAwait(false);
+                        }
+                        catch (TaskCanceledException)
+                        {
+                            // The running task has an inner delay that is likely to throw a TaskCanceledException upon token cancellation.
+                            // The task might end up leaving its main loop gracefully by chance, so we won't necessarily reach this part of
+                            // the code.
+                        }
+                        catch (Exception)
+                        {
+                            // TODO: delegate the exception handling to an Exception Callback.  Instead of delegating it to the handler,
+                            // should we surface it?
+                        }
+
+                        // Now that the task has finished, clean up what is left.  Stop and remove every partition processing task that is
+                        // still running and dispose of our dictionaries.
+
+                        await Task.WhenAll(ActivePartitionProcessors.Keys
+                            .Select(partitionId => StopPartitionProcessingIfRunningAsync(partitionId, ProcessingStoppedReason.Shutdown)))
+                            .ConfigureAwait(false);
+
+                        InstanceOwnership = null;
+                        ActivePartitionProcessors = null;
+                        ActivePartitionProcessorTokenSources = null;
+
+                        // TODO: once IsRunning is implemented, update the following comment.
+                        // We need to wait until all tasks have stopped before making the load balancing task null.  If we did it sooner, we
+                        // would have a race condition where the user could update the processing handlers while some pumps are still running.
+
+                        ActiveLoadBalancingTask = null;
+                    }
+                }
+                finally
+                {
+                    RunningTaskSemaphore.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        ///   Performs load balancing between multiple <see cref="EventProcessorClient" /> instances, claiming others' partitions to enforce
+        ///   a more equal distribution when necessary.  It also manages its own partition processing tasks and ownership.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task RunAsync(CancellationToken cancellationToken)
```

Comment:
Once we get into a stable state, outside of being release ready, I want to sit down and pull this apart to try and figure out ways to make this more deterministic and better control behavior for testing.    Maybe the week of Thanksgiving or there abouts.

Reply:
Each step is a kinda isolated, so it shouldn't be too troublesome.